From 713d4c70acc587205324cf8a0d6ab47bae5ca52b Mon Sep 17 00:00:00 2001
From: Igor Gnatenko <ignatenkobrain@fedoraproject.org>
Date: Tue, 7 Feb 2017 09:19:18 +0100
Subject: [PATCH] crypto: add support for OpenSSL 1.1

In OpenSSL 1.1:
* BN is opaque structure
  -> We'll use DH_set0_pqg() for OpenSSL 1.1 and direct access for older
* SSL is opaque structure
  -> We'll use well-known function SSL_is_server() for all OpenSSL versions

Reported-by: Vasiliy Glazov <vascom2@gmail.com>
Closes: https://github.com/eiskaltdcpp/eiskaltdcpp/issues/356
Signed-off-by: Igor Gnatenko <ignatenkobrain@fedoraproject.org>
---
 dcpp/CryptoManager.cpp | 15 +++++++++++----
 dcpp/SSLSocket.cpp     |  4 ++--
 2 files changed, 13 insertions(+), 6 deletions(-)

diff --git a/dcpp/CryptoManager.cpp b/dcpp/CryptoManager.cpp
index 08893a7..b672d92 100644
--- a/dcpp/CryptoManager.cpp
+++ b/dcpp/CryptoManager.cpp
@@ -109,10 +109,15 @@ CryptoManager::CryptoManager()
                 };
 
         if(dh) {
-            dh->p = BN_bin2bn(dh4096_p, sizeof(dh4096_p), 0);
-            dh->g = BN_bin2bn(dh4096_g, sizeof(dh4096_g), 0);
-
-            if(!dh->p || !dh->g) {
+            BIGNUM *p = BN_bin2bn(dh4096_p, sizeof(dh4096_p), 0);
+            BIGNUM *g = BN_bin2bn(dh4096_g, sizeof(dh4096_g), 0);
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+            dh->p = p;
+            dh->g = g;
+            if (!dh->p || !dh->g) {
+#else
+            if (!DH_set0_pqg(dh, p, NULL, g)) {
+#endif
                 dh.reset();
             } else {
                 SSL_CTX_set_options(serverContext, SSL_OP_SINGLE_DH_USE);
@@ -120,6 +125,8 @@ CryptoManager::CryptoManager()
                 SSL_CTX_set_tmp_dh(serverContext, (DH*)dh);
                 SSL_CTX_set_tmp_dh(serverVerContext, (DH*)dh);
             }
+            BN_free(p);
+            BN_free(g);
         }
 
         SSL_CTX_set_options(clientContext, SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3 | SSL_OP_NO_COMPRESSION);
diff --git a/dcpp/SSLSocket.cpp b/dcpp/SSLSocket.cpp
index 1e283ee..a54b7d2 100644
--- a/dcpp/SSLSocket.cpp
+++ b/dcpp/SSLSocket.cpp
@@ -54,9 +54,9 @@ bool SSLSocket::waitConnected(uint32_t millis) {
     }
 
     while(true) {
-        int ret = ssl->server?SSL_accept(ssl):SSL_connect(ssl);
+        int ret = SSL_is_server(ssl)?SSL_accept(ssl):SSL_connect(ssl);
         if(ret == 1) {
-            dcdebug("Connected to SSL server using %s as %s\n", SSL_get_cipher(ssl), ssl->server?"server":"client");
+            dcdebug("Connected to SSL server using %s as %s\n", SSL_get_cipher(ssl), SSL_is_server(ssl)?"server":"client");
             return true;
         }
         if(!waitWant(ret, millis)) {
